applications/StructuralMechanicsApplication/custom_elements/spring_damper_element.cpp

```diff
@@ -0,0 +1,601 @@
+
+    rDampingMatrix = ZeroMatrix(system_size, system_size);
+
+    if (this->Has(NODAL_DAMPING_RATIO) || this->Has(NODAL_ROTATIONAL_DAMPING_RATIO)) {
+        array_1d<double, msLocalSize> elemental_damping_ratio = ZeroVector(msLocalSize);
+        if (this->Has(NODAL_DAMPING_RATIO)) {
+            const array_1d<double, 3>& nodal_damping = this->GetValue(NODAL_DAMPING_RATIO);
+
+            elemental_damping_ratio[0] = nodal_damping[0];
+            elemental_damping_ratio[1] = nodal_damping[1];
+
+            if constexpr (TDim == 3) {
+                elemental_damping_ratio[2] = nodal_damping[2];
+            }
+        }
+
+        if (this->Has(NODAL_ROTATIONAL_DAMPING_RATIO)) {
+            const array_1d<double, 3>& nodal_rotational_damping = this->GetValue(NODAL_ROTATIONAL_DAMPING_RATIO);
+            if constexpr (TDim == 2) {
+                elemental_damping_ratio[2] = nodal_rotational_damping[2];
+            }
+            else if constexpr (TDim == 3) {
+                elemental_damping_ratio[3] = nodal_rotational_damping[0];
+                elemental_damping_ratio[4] = nodal_rotational_damping[1];
+                elemental_damping_ratio[5] = nodal_rotational_damping[2];
+            }
+        }
+
+        for (std::size_t i = 0; i < msLocalSize; ++i) {
+            rDampingMatrix(i, i) += elemental_damping_ratio[i];
+            rDampingMatrix(i + msLocalSize, i + msLocalSize) += elemental_damping_ratio[i];
+            rDampingMatrix(i, i + msLocalSize) -= elemental_damping_ratio[i];
+            rDampingMatrix(i + msLocalSize, i) -= elemental_damping_ratio[i];
+        }
+    }
+
+    KRATOS_CATCH("");
+}
+
+
+template<std::size_t TDim>
+void SpringDamperElement<TDim>::ConstCalculateLocalSystem(MatrixType& rLeftHandSideMatrix, VectorType& rRightHandSideVector) const
+{
+    KRATOS_TRY;
+
+    /* Calculate elemental system */
+    // Compute RHS (RHS = rRightHandSideVector = Fext - Fint)
+    this->ConstCalculateRightHandSide(rRightHandSideVector);
+
+    // Compute LHS
+    this->ConstCalculateLeftHandSide(rLeftHandSideMatrix);
+
+    KRATOS_CATCH("");
+}
+
+
+template<std::size_t TDim>
+void SpringDamperElement<TDim>::ConstCalculateLeftHandSide(MatrixType& rLeftHandSideMatrix) const
+{
+    KRATOS_TRY;
+    // Resizing the LHS
+    std::size_t system_size = msElementSize;
+
+    if (rLeftHandSideMatrix.size1() != system_size) {
+        rLeftHandSideMatrix.resize(system_size, system_size, false);
+    }
+
+    noalias(rLeftHandSideMatrix) = ZeroMatrix(system_size, system_size); //resetting LHS
+
+    // elemental_stiffness: kx, ky, kz, cpx, cpy, cpz
+    array_1d<double, msLocalSize > elemental_stiffness = ZeroVector(msLocalSize);
+    array_1d<double, 3> nodal_stiffness = ZeroVector(3);
+    array_1d<double, 3> nodal_rot_stiffness = ZeroVector(3);
+
+    if (this->Has(NODAL_DISPLACEMENT_STIFFNESS))
+    {
+        nodal_stiffness = this->GetValue(NODAL_DISPLACEMENT_STIFFNESS);
+    }
```

Comment:
```suggestion
    const array_1d<double, 3>& nodal_stiffness = this->GetValue(NODAL_DISPLACEMENT_STIFFNESS);
```

this check is already done in the `DataValueContainer` (but only in the `const` version of the `GetValue` function)
https://github.com/KratosMultiphysics/Kratos/blob/0fb273caebbae77e5645207a731a057349be9719/kratos/containers/data_value_container.h#L206-L214

this one one of the main reason/advantage of having the `const` functions

Reply:
aha now I understand